### PR TITLE
Deferred syncfs on file changes

### PR DESCRIPTION
**Pull request by timdream**
This implements #3453 behind a build config named `IDBFS_DEFER_WRITE`. Two tests are created to ensure syncfs triggers in various situations.

I am not really happy about the naming (`IDBFS_DEFER_WRITE`). I would love to change it if there is something better!

After this gets merged, I would like to patch the document of IDBFS to explain the features (this one and the previous `fsync` one.)

Thanks for you help!

### CODE REVIEW

File: tests/fs/test_idbfs_defer_write_on_exit.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+int result = 1;
+
+int main() {
+  struct stat st;
+  int fd;
+
+  // We first open and close an empty file in the directory and leave, which
+  // should trigger syncfs() on exit. We will then verify the existance of the
+  // file on the !FIRST run.
```
Comment (timdream) [inline tests/fs/test_idbfs_defer_write_on_exit.c]: BTW this test also rely on the fact syncfs can't possibly spend for more than 1 sec (the time between we `REPORT_RESULT();` and we call `window.close()` in `tests/runner.py`).

File: src/library_idbfs.js [hunk 2]
```diff
@@ -15,16 +15,56 @@ mergeInto(LibraryManager.library, {
...
+#if IDBFS_DEFER_WRITE
+    DEFER_WRITE_TIMEOUT: 10000,
+    deferredSyncFs: function(mount) {
+      // If there is already a timer, don't set another one.
+      if (mount.mountSyncfsTimer) {
+        return;
+      }
+
+      // Set a timer what would trigger syncfs after 10 sec.
+      mount.mountSyncfsTimer = setTimeout(function() {
```
Comment (kripken) [inline src/library_idbfs.js]: Best to use `Browser.safeSetTimeout`, which will avoid running the timeout if the application exited already.

File: src/library_idbfs.js [hunk 3]
```diff
@@ -15,16 +15,56 @@ mergeInto(LibraryManager.library, {
       // reuse all of the core MEMFS functionality
       return MEMFS.mount.apply(null, arguments);
     },
+#if IDBFS_DEFER_WRITE
+    DEFER_WRITE_TIMEOUT: 10000,
+    deferredSyncFs: function(mount) {
+      // If there is already a timer, don't set another one.
+      if (mount.mountSyncfsTimer) {
```
Comment (kripken) [inline src/library_idbfs.js]: can put the return on the same line for short lines like this

File: src/library_idbfs.js [hunk 4]
```diff
@@ -15,16 +15,56 @@ mergeInto(LibraryManager.library, {
...
+          }
+        });
+      }, IDBFS.DEFER_WRITE_TIMEOUT);
+
+      // Also set an atExit hook (if we have not yet done so.)
+      if (!mount.atExitSyncFsHookAdded) {
+        mount.atExitSyncFsHookAdded = true;
+        Module['addOnExit'](function() {
+          // We might not need to syncfs() at this time ...
+          if (!mount.mountSyncfsTimer) {
```
Comment (kripken) [inline src/library_idbfs.js]: same as above

File: src/library_idbfs.js [hunk 5]
```diff
@@ -15,16 +15,56 @@ mergeInto(LibraryManager.library, {
...
+              console.log("IDBFS.syncfs threw an exception: ", err);
+            }
+          });
+        });
+      }
+    },
+#endif
     syncfs: function(mount, populate, callback) {
-      IDBFS.getLocalSet(mount, function(err, local) {
+      IDBFS.getRemoteSet(mount, function(err, remote) {
```
Comment (kripken) [inline src/library_idbfs.js]: why were these flipped?

File: src/library_idbfs.js (hunk 2 above)
Comment (kripken) [inline src/library_idbfs.js]: but need to make sure that can work with clearTimeout, I'm not sure offhand

File: src/settings.js [hunk 6]
```diff
@@ -480,7 +480,7 @@ var EMTERPRETIFY_ADVISE = 0; // Performs a static analysis to suggest which func
                              // all things reachable from that function, as they may need to be in the YIELDLIST as well).
                              // Note that this depends on things like inlining. If you run this with different inlining than
                              // when you use the list, it might not work.
-
+var IDBFS_DEFER_WRITE = 0; // Set a timer to write into IndexedDB when file changes.
```
Comment (kripken) [inline src/settings.js]: how about `IDBFS_AUTOMATIC_SYNC`, since this automatic syncing also happens on exit?

**Comment by kripken** [comment]
First thing, sorry for the delay in reviewing.

Ok, this looks quite good :) I wrote some comments above. The only significant issue I am unsure about is the testing. I didn't quite understand how the first test works - how does it verify what it says it tests, is exactly what happens? For the second, I worry about the delay of `REPORT_RESULT` making this a little unstable.

How about testing like this, though: Do some file operations, then do an `emscripten_sleep` (using `EMTERPRETIFY_ASYNC`) for a few  seconds, and check if it was actually written out. If not, wait some more, up to some timeout that fails. I'm not sure how to do the check, though - perhaps we could mount it again at another location, and see if that gets synced?

File: src/library_idbfs.js (hunk 5 above)
Comment (timdream) [inline src/library_idbfs.js]: I flip this to ensure we are cleaning the timeout at the same function loop of `IDBFS.getLocalSet()`.

Comment (kripken) [inline src/library_idbfs.js]: i'm not sure why that matters? couldn't the clearing be done in the very first line of `syncfs`?

**Comment by timdream** [comment]
I might not be available to work on this in coming weeks due to things happen in the company. Will come back to this soon.

**Comment by timdream** [comment]
Rebased and addressed the non-testing part. Please come back after I have fixed the tests.

**Comment by timdream** [comment]
I am very sorry for the delay -- will get to this eventually. The main hold up currently is to think of a way to properly test the feature -- I don't think IDBFS allow me to remount the same IndexedDB database on a different virtual file system path -- but maybe I am wrong on this and will spend time to look into it.

**Comment by santeriv** [comment]
Any updates on this?

Conflicting files: src/settings.js (only new flag?)

**Comment by timdream** [comment]
@santeriv Sorry, I am no longer actively working on it. The last comment still stands, feel free to take it over.

I would also like to mention that people seem to implement their own FS in various projects, so the need of adding the capability to IDBFS in the Emscripten core is perhaps less desired.

**Comment by santeriv** [comment]
@timdream Thanks for fast response and good thoughts. I cannot commit to this one.

 Yet as a user of em-sdk, I'd hope enabling auto sync using IDBFS would be simpler - using em-sdk (custom build), em-fs-plugins?!. Just thoughts as well.

**Pull request closed without merging**